File: sdk/runanywhere-react-native/packages/core/src/Features/VoiceAgent/VoiceAgentMicDriver.ts
/**
 * VoiceAgentMicDriver.ts
 *
 * Audio ingress for the voice agent. The C ABI owns NO microphone access
 * (rac_voice_agent.h "Audio-Ingress Contract"): the platform SDK must capture
 * mic audio and push complete utterances into the C core, or the session is
 * dead-air. This driver implements ingress mode 1 (per-utterance turns):
 * capture 16 kHz mono PCM16 via {@link AudioCaptureManager}, segment utterances
 * with energy-based endpointing, and submit each utterance through
 * `rac_voice_agent_process_voice_turn_proto` (`voiceAgentProcessTurnProto`).
 * Turn VoiceEvents fan out to the handle callback, so
 * `RunAnywhere.streamVoiceAgent()` collectors observe them without extra wiring.
 *
 * Mirrors `sdk/runanywhere-swift/.../VoiceAgentMicDriver.swift` and
 * `sdk/runanywhere-kotlin/.../VoiceAgentMicDriver.kt`. Endpointing is
 * energy-based; mic chunks that arrive while a turn is processing are discarded
 * (strict turn-taking, no barge-in — also avoids transcribing the device's own
 * TTS output).
 */

import { AudioCaptureManager } from '../VoiceSession/AudioCaptureManager';
import { AudioPlaybackManager } from '../VoiceSession/AudioPlaybackManager';
import { SDKLogger } from '../../Foundation/Logging/Logger/SDKLogger';
import { requireNativeModule } from '../../native';
import { arrayBufferToBytes } from '../../services/ProtoBytes';
import { VoiceAgentResult as VoiceAgentResultMessage } from '@runanywhere/proto-ts/voice_agent_service';

const SAMPLE_RATE_HZ = 16_000;
const BYTES_PER_SAMPLE = 2;
/** Absolute floor for the adaptive speech threshold (normalized RMS). */
const SPEECH_RMS_THRESHOLD = 0.015;
/** Speech must exceed this multiple of the tracked ambient noise floor. */
const SPEECH_FLOOR_MULTIPLIER = 2.2;
/** Per-chunk rate at which the ambient floor creeps up toward louder ambient. */
const NOISE_FLOOR_RISE = 0.05;
/** Trailing silence that closes an utterance. */
const END_OF_UTTERANCE_SILENCE_MS = 800;
/** Utterances with less accumulated speech than this are noise. */
const MIN_SPEECH_MS = 300;
/** Hard cap so a noisy room cannot grow an unbounded buffer. */
const MAX_UTTERANCE_MS = 15_000;
/** Leading chunks kept so the utterance onset is not clipped. */
const PRE_ROLL_CHUNKS = 3;

/**
 * Captures mic audio and drives per-utterance voice-agent turns. {@link start}
 * runs until {@link stop} is called. Segmentation runs synchronously inside the
 * capture callback; an utterance end kicks off an async turn, during which
 * incoming chunks are dropped (the `processing` gate) so the mic stays gated
 * while the device thinks / speaks.
 */
export class VoiceAgentMicDriver {
  private readonly logger = new SDKLogger('VoiceAgentMic');
  private readonly capture = new AudioCaptureManager();
  private readonly playback = new AudioPlaybackManager();

  private stopped = false;
  private processing = false;

  // Segmentation state.
  private preRoll: Uint8Array[] = [];
  private utterance: Uint8Array[] = [];
  private inSpeech = false;
  private speechMs = 0;
  private silenceMs = 0;
  private noiseFloor = SPEECH_RMS_THRESHOLD;

  /** Begin mic capture + segmentation. Resolves once capture has started. */
  async start(): Promise<void> {
    const granted = await this.capture.requestPermission();
    if (!granted) {
      throw new Error('Microphone permission denied');
    }
    this.stopped = false;
    // The voice agent runs a single full-duplex (.playAndRecord) session for the
    // whole turn-taking loop so TTS replies can play through the speaker while the
    // mic stays live. Activate it BEFORE startRecording so capture reuses it
    // instead of forcing the output-only .record session (which silences replies
    // and trips cannotStartPlaying). Mirrors the iOS Swift driver's
    // configureVoiceAudioSession(); no-op on Android.
    await this.capture.activateAudioSession();
    await this.capture.startRecording((chunk) => this.onChunk(chunk));
    this.logger.info('Voice-agent mic capture started');
  }

  /** Stop capture + playback and reset segmentation state. */
  async stop(): Promise<void> {
    if (this.stopped) return;
    this.stopped = true;
    try {
      this.capture.stopRecording();
    } catch {
      /* noop */
    }
    try {
      this.playback.stop();
    } catch {
      /* noop */
    }
    this.preRoll = [];
    this.utterance = [];
    this.inSpeech = false;
    this.logger.info('Voice-agent mic capture stopped');
  }

  private onChunk(chunk: Uint8Array): void {
    // Drop chunks while a turn is in flight (no barge-in) or after stop().
    if (this.stopped || this.processing || chunk.byteLength === 0) return;

    const chunkMs =
      (chunk.byteLength * 1000) / (SAMPLE_RATE_HZ * BYTES_PER_SAMPLE);
    // Adaptive endpointing: a fixed RMS threshold misses the end-of-utterance
    // pause on devices whose mic noise floor sits above the constant. Track the
    // ambient floor — drop instantly to any quieter level, creep up only while
    // not in speech — and require a chunk to rise clearly above it.
    const level = VoiceAgentMicDriver.rms(chunk);
    const speechThreshold = Math.max(
      SPEECH_RMS_THRESHOLD,
      this.noiseFloor * SPEECH_FLOOR_MULTIPLIER
    );
    const isSpeech = level >= speechThreshold;
    if (!this.inSpeech) {
      if (level < this.noiseFloor) {
        this.noiseFloor = level;
      } else if (!isSpeech) {
        this.noiseFloor += (level - this.noiseFloor) * NOISE_FLOOR_RISE;
      }
    }

    if (!this.inSpeech) {
      this.preRoll.push(chunk);
      if (this.preRoll.length > PRE_ROLL_CHUNKS) {
        this.preRoll.shift();
      }
      if (isSpeech) {
        this.inSpeech = true;
        this.speechMs = chunkMs;
        this.silenceMs = 0;
        this.utterance = [...this.preRoll];
        this.preRoll = [];
      }
      return;
    }

    this.utterance.push(chunk);
    if (isSpeech) {
      this.speechMs += chunkMs;
      this.silenceMs = 0;
    } else {
      this.silenceMs += chunkMs;
    }

    const utteranceBytes = this.utterance.reduce((n, c) => n + c.byteLength, 0);
    const utteranceMs =
      (utteranceBytes * 1000) / (SAMPLE_RATE_HZ * BYTES_PER_SAMPLE);
    if (
      this.silenceMs >= END_OF_UTTERANCE_SILENCE_MS ||
      utteranceMs >= MAX_UTTERANCE_MS
    ) {
      const speechMs = this.speechMs;
      const audio = VoiceAgentMicDriver.concat(this.utterance);
      this.inSpeech = false;
      this.utterance = [];
      this.speechMs = 0;
      this.silenceMs = 0;
      if (speechMs >= MIN_SPEECH_MS) {
        // Gate the mic for the duration of the turn (and TTS playback) so the
        // device does not transcribe its own reply. `processing` is cleared
        // once the turn + playback finish.
        this.processing = true;
        void this.processTurn(audio).finally(() => {
          this.processing = false;
        });
      } else {
        this.logger.debug(
          `Utterance discarded (${Math.round(speechMs)}ms speech < ${MIN_SPEECH_MS}ms)`
        );
      }
    }
  }

  private async processTurn(audio: Uint8Array): Promise<void> {
    if (this.stopped || audio.byteLength === 0) return;
    this.logger.info(`Submitting voice turn (${audio.byteLength} bytes)`);
    try {
      const native = requireNativeModule();
      const resultBytes = await native.voiceAgentProcessTurnProto(
        VoiceAgentMicDriver.toArrayBuffer(audio)
      );
      if (this.stopped) return;
      const bytes = arrayBufferToBytes(resultBytes);
      if (bytes.byteLength === 0) {
        this.logger.warning('Voice turn returned an empty result');
        return;
      }
      const result = VoiceAgentResultMessage.decode(bytes);
      if (result.errorMessage && result.errorMessage.length > 0) {
        this.logger.warning(`Voice turn failed: ${result.errorMessage}`);
      }
      await this.playReply(result);
    } catch (error) {
      this.logger.warning(
        `Voice turn threw: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Play the turn's synthesized reply through the shared playback sink. Runs
   * while `processing` is still set, so the mic stays gated while the device
   * speaks (no self-transcription). Commons returns `synthesized_audio` as a
   * complete WAV (`rac_audio_float32_to_wav`), so the bytes are handed to the
   * native player as-is — not re-encoded.
   */
  private async playReply(
    result: ReturnType<typeof VoiceAgentResultMessage.decode>
  ): Promise<void> {
    const audio = result.synthesizedAudio;
    if (!audio || audio.byteLength === 0) return;

    const wav = VoiceAgentMicDriver.toArrayBuffer(audio);
    this.logger.info(`Playing agent reply (${audio.byteLength} WAV bytes)`);
    try {
      await this.playback.playWav(wav);
    } catch (error) {
      this.logger.warning(
        `Agent reply playback failed: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /** Normalized RMS of a 16 kHz mono Int16-LE chunk (0..1). */
  private static rms(chunk: Uint8Array): number {
    const sampleCount = Math.floor(chunk.byteLength / BYTES_PER_SAMPLE);
    if (sampleCount === 0) return 0;
    const view = new DataView(
      chunk.buffer,
      chunk.byteOffset,
      sampleCount * BYTES_PER_SAMPLE
    );
    let sum = 0;
    for (let i = 0; i < sampleCount; i++) {
      const sample = view.getInt16(i * BYTES_PER_SAMPLE, true);
      sum += sample * sample;
    }
    return Math.sqrt(sum / sampleCount) / 32767;
  }

  private static concat(chunks: Uint8Array[]): Uint8Array {
    const total = chunks.reduce((n, c) => n + c.byteLength, 0);
    const out = new Uint8Array(total);
    let offset = 0;
    for (const c of chunks) {
      out.set(c, offset);
      offset += c.byteLength;
    }
    return out;
  }

  private static toArrayBuffer(bytes: Uint8Array): ArrayBuffer {
    return bytes.buffer.slice(
      bytes.byteOffset,
      bytes.byteOffset + bytes.byteLength
    ) as ArrayBuffer;
  }
}

File: sdk/runanywhere-flutter/packages/runanywhere/lib/features/voice_agent/services/voice_agent_mic_driver.dart
// SPDX-License-Identifier: Apache-2.0
//
// voice_agent_mic_driver.dart — audio ingress for the Flutter voice agent.
//
// The C ABI owns NO microphone (rac_voice_agent.h "Audio-Ingress Contract"):
// the platform SDK must capture mic audio and push raw frames into the C core
// via `rac_voice_agent_feed_audio_proto`, or the session is dead air.
//
// Mirrors Kotlin/Swift `VoiceAgentMicDriver`: capture 16 kHz mono PCM16 via
// [AudioCaptureManager], feed every chunk to commons (which owns energy VAD,
// hangover, pre-roll, and the STT → LLM → TTS turn), forward VoiceEvents from
// [VoiceAgentStreamAdapter], and play the synthesized WAV reply returned
// inline. NO SDK-side RMS / segmenter / endpointing.

import 'dart:async';
import 'dart:collection';
import 'dart:typed_data';

import 'package:runanywhere/adapters/voice_agent_stream_adapter.dart';
import 'package:runanywhere/features/stt/services/audio_capture_manager.dart';
import 'package:runanywhere/features/tts/services/audio_playback_manager.dart';
import 'package:runanywhere/foundation/logging/sdk_logger.dart';
import 'package:runanywhere/generated/model_types.pbenum.dart' as model_pb;
import 'package:runanywhere/generated/ra_defaults_pool.dart';
import 'package:runanywhere/generated/ra_result_codes.dart';
import 'package:runanywhere/generated/voice_agent_service.pb.dart'
    as voice_agent_pb;
import 'package:runanywhere/generated/voice_events.pb.dart' as voice_events_pb;
import 'package:runanywhere/native/dart_bridge.dart';
import 'package:runanywhere/native/types/basic_types.dart';

/// Captures mic audio and feeds raw frames to the in-core voice agent. [start]
/// begins capture; [events] streams every turn's VoiceEvents; [stop] tears
/// capture + playback down. Mirrors Kotlin/Swift `VoiceAgentMicDriver`.
class VoiceAgentMicDriver {
  VoiceAgentMicDriver();

  static final _logger = SDKLogger('VoiceAgentMic');

  static const int _sampleRateHz = RADefaultsAudioCapture.micSampleRateHz;
  static const int _channelCapacity =
      RADefaultsAudioCapture.micChannelCapacity;
  static const Duration _feedIdleSleep = Duration(milliseconds: 20);

  final AudioCaptureManager _capture = AudioCaptureManager();
  final AudioPlaybackManager _playback = AudioPlaybackManager();
  final StreamController<voice_events_pb.VoiceEvent> _out =
      StreamController<voice_events_pb.VoiceEvent>();
  final Queue<Uint8List> _queue = Queue<Uint8List>();

  StreamSubscription<Uint8List>? _micSub;
  StreamSubscription<voice_events_pb.VoiceEvent>? _eventSub;
  bool _stopped = false;
  bool _feedRunning = false;

  /// VoiceEvents produced by each turn (userSaid, llm tokens, audio, pipeline
  /// state). The public eventStream yields from this. Sourced from the
  /// commons proto callback via [VoiceAgentStreamAdapter] — not from a local
  /// segmenter driving `process_turn_proto`.
  Stream<voice_events_pb.VoiceEvent> get events => _out.stream;

  /// Begin mic capture + feed loop. On permission/capture failure the [events]
  /// stream is closed with an error so the collector can surface it.
  Future<void> start() async {
    // The voice agent runs a single full-duplex (.playAndRecord) session for the
    // whole turn-taking loop — the `record` plugin configures it on
    // startRecording (defaultToSpeaker by default). Playback must NOT switch the
    // session to the output-only `.playback` category, or it trips
    // AVAudioSessionErrorInsufficientPriority ('!pri', OSStatus 561017449) and
    // the reply is dropped. Mirrors the iOS Swift driver
    // (playback.managesAudioSession = false).
    _playback.managesAudioSession = false;
    _stopped = false;

    // Attach the proto callback BEFORE capture so events from the first turn
    // are not dropped (mirrors Swift VoiceSession.events).
    final handle = await DartBridge.voiceAgent.getHandle();
    _eventSub = VoiceAgentStreamAdapter(handle).stream().listen(
      (event) {
        if (!_out.isClosed) _out.add(event);
      },
      onError: (Object e, StackTrace st) {
        if (!_out.isClosed) _out.addError(e, st);
      },
      cancelOnError: false,
    );

    final stream =
        await _capture.startRecording(sampleRate: _sampleRateHz, numChannels: 1);
    if (stream == null) {
      await _eventSub?.cancel();
      _eventSub = null;
      _out.addError(StateError(
          'Microphone capture unavailable (permission denied or busy)'));
      unawaited(_out.close());
      return;
    }
    _logger.info('Voice-agent mic capture started');
    _micSub = stream.listen(
      _enqueueChunk,
      onError: (Object e, StackTrace st) => _logger.warning('Mic error: $e'),
      cancelOnError: false,
    );
    unawaited(_feedLoop());
  }

  /// Stop capture + playback and close the event stream.
  Future<void> stop() async {
    if (_stopped) return;
    _stopped = true;
    await _micSub?.cancel();
    _micSub = null;
    await _capture.stopRecording();
    await _playback.stop();
    await _eventSub?.cancel();
    _eventSub = null;
    _queue.clear();
    // Wait for the feed loop to observe _stopped before closing events.
    while (_feedRunning) {
      await Future<void>.delayed(_feedIdleSleep);
    }
    if (!_out.isClosed) {
      await _out.close();
    }
    _logger.info('Voice-agent mic capture stopped');
  }

  void _enqueueChunk(Uint8List chunk) {
    if (_stopped || chunk.isEmpty) return;
    _queue.add(chunk);
    while (_queue.length > _channelCapacity) {
      _queue.removeFirst();
    }
  }

  List<Uint8List> _drainChunks() {
    if (_queue.isEmpty) return const <Uint8List>[];
    final drained = _queue.toList(growable: false);
    _queue.clear();
    return drained;
  }

  /// Drains captured frames and feeds them to commons. The core blocks the
  /// feed call for the duration of a turn when an utterance closes and returns
  /// the synthesized reply inline as WAV. Per-stage VoiceEvents fan out through
  /// the handle callback (forwarded by [_eventSub]).
  Future<void> _feedLoop() async {
    _feedRunning = true;
    try {
      while (!_stopped) {
        final chunks = _drainChunks();
        if (chunks.isEmpty) {
          await Future<void>.delayed(_feedIdleSleep);
          continue;
        }

        for (final chunk in chunks) {
          if (_stopped) return;

          try {
            final frame = voice_agent_pb.VoiceAgentAudioFrame(
              audioData: chunk,
              sampleRateHz: _sampleRateHz,
              channels: 1,
              encoding: model_pb.AudioEncoding.AUDIO_ENCODING_PCM_S16_LE,
              isFinal: false,
            );
            final outcome = await DartBridge.voiceAgent.feedAudioProto(frame);
            if (_stopped) return;

            if (outcome.status == RacResultCodes.errorNotInitialized) {
              throw StateError('Voice agent is no longer initialized');
            }
            if (outcome.status != RAC_SUCCESS) {
              _logger.warning('Voice feed failed: rc=${outcome.status}');
              _queue.clear();
              continue;
            }

            final reply = outcome.result?.synthesizedAudio;
            if (reply != null && reply.isNotEmpty) {
              _logger.info('Playing agent reply (${reply.length} WAV bytes)');
              // Frames captured while the turn was computed predate playout —
              // drop them so the reply onset cannot seed the echo estimate
              // (mirrors Swift/Kotlin).
              _queue.clear();
              await _playReply(
                reply is Uint8List ? reply : Uint8List.fromList(reply),
              );
            }
          } catch (e) {
            _logger.warning('Voice feed failed: $e');
            _queue.clear();
            if (!_out.isClosed) {
              _out.addError(e);
            }
          }
        }
      }
    } finally {
      _feedRunning = false;
    }
  }

  Future<void> _playReply(Uint8List wav) async {
    if (wav.isEmpty || _stopped) return;
    try {
      await _playback.play(wav);
    } catch (e) {
      _logger.warning('Agent reply playback failed: $e');
    }
  }
}
